### ISSUE TITLE
AGS 4: Parent class for game objects

### ISSUE THREAD

**Issue by ivan-mogilko** [adventuregamestudio/ags#1228]
As we expand script api and try to bring all game objects to consistent behavior it is clear that they share functionality but do not have any base class that would help apply it uniformly. For example, if one likes to add a new generic visual property, they would have to add it separately to each of the game object types. This includes both script declaration and api handlers in the engine code.

(The only example of a parent class in existing script API is GUIControl that shares functionality with rest of control classes)

The proposal is to add one or more base classes to the script api, move existing shared properties and functions there, and use these in future whenever we might need to add more api shared among all objects.

I think we may begin with most obvious:
* properties defining object location;
* properties defining generic looks;

In the past I wrote a wiki page with some [thoughts on potential class hierarchy](https://github.com/adventuregamestudio/ags/wiki/AGS-4-Draft:-Game-Data-and-Entities#---default-entity-class-hierarchy-draft--example). That is not a real plan, just something I was thinking over.

My suggestion is to have at least two parent classes for now: I called them "Thing" ("Thing2D" is an alternate proposal) and "GraphicItem".

**Thing** or **Thing2D** defines an object in 2D space, which basically has a position (X, Y). I believed it should be a separate base class, because there might be subclasses that need only that. An example could be an "audio emitter" that plays sound simulating it coming from certain position in space.
Later this class may be amended with settings, ti give some example, which define which coordinate space (layer) the object is located in.

**GraphicItem** defines an object which has a graphic representation. Note that it does not have to have explicit Graphic (sprite) property, as it is not necessarily drawing one of the game sprites (see GUI). What it should have is a full collection of settings that adjust looks: visibility, transparency, tint, blend mode (one of the recent additions), and so forth.
Scaling and rotation should go there too; as anything we add for visual effect (shader support?).

Following existing objects will become a child of GraphicItem: Character, GUI, GUIControl, Object, Overlay.

Viewport and Camera are a questionable at the moment.
It's not clear to me at the moment whether Viewport should be a child of Thing or GraphicItem. It depends on how do we see the viewport as an object. Technically it's hardcoded as the rectangle where camera is rendered. It may be made visible, and I was thinking about adding transparency. Probably it could have blend mode too, as it renders, maybe be even e.g. rotated.
However, it's scaling only makes sense in connection with camera (it scales whatever camera shows). If we suppose for the moment that camera's image for viewport is like sprite to an object, then yes, there's definitely same kind of scaling.
Maybe it's functionality could be reviewed to be represented more consistently.
Hypothetically in the future AGS may have camera applied to any object replacing its graphic, so this is also something to keep in mind.

OTOH, Camera may be rotated, and probably could be applied a tint or another shader, but transparency and blend mode would not make much sense there, as it provides certain image but does not apply that image over anything.
Even Visible flag does not seem to make much sense... or does it? e.g. one could set camera visible = false, then all viewports that connect to it will display nothing.
It's also probably possible to have it originate from GraphicItem, but ignore certain properties as non-applicable. If that makes sense.

**Comment by ericoporto** [comment]
Tween and any other module that extends existing game objects I think could benefit greatly from this change, since there's a lot of code duplication today to add things like TweenPosition.

**Comment by ericoporto** [comment]
I have been thinking about this, and an alternative would be to have a Transform object, which could be modified with Scale, Translation and Rotation properties, and have the game objects have an attribute to their transform objects, some other object to map pixel transformations (tint, blend, ...) and then the game objects would own these. About transform, the origin is different (characters have at their feet, room objects at bottom left, GUIs at top left), so this should be taken account somehow.

**Comment by AlanDrake** [comment]
Having a Transform property was actually in the plans. The doubts were if it should only be visual or more.

**Comment by ivan-mogilko** [comment]
@ericoporto 
> and an alternative would be to have a Transform object

Could you clarify, you suggest this as an alternative to _what_ (there are many proposals mentioned above)?

**Comment by ivan-mogilko** [comment]
I realized that forgot to mention another important use for having a class hierarchy of the script objects: it's the ability to cast types. I.e. if you only need to work with basic object's properties or functions, you would do something like (pseudo-code):
<pre>
BaseType* obj = BaseType.GetAtScreenXY(mouse.x, mouse.y);
obj.RunInteraction(eModeLookat);
</pre>
Right now you have to check for location type, or call GetAtScreenXY repeatedly for each possible object type.

**Comment by ericoporto** [comment]
> @ericoporto
> 
> > and an alternative would be to have a Transform object
> 
> Could you clarify, you suggest this as an alternative to _what_ (there are many proposals mentioned above)?

I meant similar to Unity Transform usage: https://docs.unity3d.com/ScriptReference/Transform.html

It would be an attribute of script objects and group the behavior of things that either Translate, Rotate or Resize the object - the matrix transformations. I would not use it to hold positional information since that is used for game logic too, but instead the Translation I was thinking was more in the sense of an offset, like the currently z property which I think is a bad name for YOffset. And perhaps Offset is a less confusing name (vs Translate as from languages).

**Comment by ivan-mogilko** [comment]
Yes, Unity3d is component-based, while AGS script is rather hierarchy based, so unless we change to component-based system, these attributes will have to be inherited from some top parent or mid-parent type, and not added to every type that has to have them (like you do in unity3d an similar engines).

**Comment by ivan-mogilko** [comment]
A *very dirty* experimental branch:
https://github.com/ivan-mogilko/ags-refactoring/tree/experiment--baseclassinscript

Majority of changes are purposed to separate few classes into "data class" that is loaded from file (and shared between engine, editor and tools), and "runtime class", which exists strictly in the engine. Runtime class must inherit a "script entity" parent class. This is done for: Character and ScriptAudioClip, because these are exported to script directly. Also has to be done for GUI Control class, but I did not do that in that branch yet (wanted to test that the idea works in general first).

**Comment by ericoporto** [comment]
Uhm, to be honest in my mind I completely ignored everything else, as my first thought was "can this make tween implementation smaller?", and other than this the other thing bugging me is how the coordinates of the thing map to the screen - are we room or screen coordinates? What is the origin (bottom middle, top left, bottom left)? I guess these things are necessary.

The other completely unrelated to this approach would be everything owns an overlay (room or screen) and you get to access them.

Overall from my actual initial look to this was with the goal of making some script modules easier to code.

I guess the other look at this is to reduce the amount of effort to add new functionality to things in the engine - assuming it can be done by just adding it to whatever the parent object. I don't know how true this will be. Is the idea here to later apply the same to the C# data counterparts?

**Comment by ivan-mogilko** [comment]
> The other completely unrelated to this approach would be everything owns an overlay (room or screen) and you get to access them.

There was a discussion on AGS Discord recently that reminded me of this topic; and it came to me that this might be a better approach rather than make everything inherit some sort of "drawable object" class.
Not only because it's sometimes difficult to put everything in a single hierarchy, but also because some objects have complex logic, where same kind of property may mean more than it means in a sprite.

Good example is a Character class. Character has Scaling, but besides only scaling the sprite, character's scaling affects a number of other things, such as walking speed and footstep sound volume. Then, Character has Rotation, but, as I mentioned back when implementing rotation in AGS 4.0, there are two kinds of rotations: graphical and logical. Graphical rotation affects the sprite look only, and logical affects the object's point of interaction with other things, such as room. I.e. you may imagine that "truly" rotated character is walking where its feet are *after rotation*. Right now that's not the case; that's also the reason why I called the property GraphicRotation instead of simply Rotation.

This consideration brings me to a thought that we might actually go the "component" way, and have game objects have some sort of a "Sprite" property that returns a reference to a drawable item. This may be an Overlay, or a class of another name, that's a matter of further debate. But then we may have 2 hierarchies: a hierarchy of a game object, something that has a script name and ID, and interactions, and so forth, and a hierarchy of pure graphic objects (this may be a small hierarchy, of just couple types).

It would be great if we had this in the first 4.0 version, but it's far too late for that. This is something that must be thought through well, and planned with all the nuances, noting everything that will be affected. As such sort of change should be done in one batch.
